Userland/Libraries/LibWeb/HTML/FormAssociatedElement.cpp

```diff
@@ -176,6 +177,29 @@ String FormAssociatedElement::relevant_value() const
     VERIFY_NOT_REACHED();
 }
 
+WebIDL::ExceptionOr<void> FormAssociatedElement::set_relevant_value(String const& new_value)
+{
+    auto& html_element = form_associated_element_to_html_element();
+    if (is<HTMLInputElement>(html_element))
+        TRY(static_cast<HTMLInputElement&>(html_element).set_value(new_value));
+    else if (is<HTMLTextAreaElement>(html_element))
+        static_cast<HTMLTextAreaElement&>(html_element).set_value(new_value);
+    else
+        VERIFY_NOT_REACHED();
+    return {};
+}
+
+void FormAssociatedElement::set_dirty_value_flag(bool flag)
+{
+    auto& html_element = form_associated_element_to_html_element();
+    if (is<HTMLInputElement>(html_element))
+        static_cast<HTMLInputElement&>(html_element).set_dirty_value_flag({}, flag);
+    else if (is<HTMLTextAreaElement>(html_element))
+        static_cast<HTMLTextAreaElement&>(html_element).set_dirty_value_flag({}, flag);
+    else
+        VERIFY_NOT_REACHED();
+}
```

Comment:
Would it not be more fitting to implement these as virtual functions, like much of the rest of `FormAssociatedElement`, instead of manually branching on the type of the element?

Reply:
I've updated the PR to avoid this branching by moving the selection API methods into their own subclass - `FormAssociatedTextControlElement`. This contains things only relevant to text controls (`HTMLInputElement` and `HTMLTextAreaElement`). More stuff could probably be moved there - like the dirty value flag, but I've tried to keep it as simple as possible for now.